### DISCUSSION TITLE
[Bug]: MockServerContainer Doesn't Get Reused

### DISCUSSION THREAD

**Discussion by akefirad** [testcontainers/testcontainers-java#6085] [Q&A]
### Module

MockServer

### Testcontainers version

1.17.5

### Using the latest Testcontainers version?

Yes

### Host OS

macOS

### Host Arch

x86

### Docker version

```shell
Client:
 Cloud integration: v1.0.22
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:46:56 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:56 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### What happened?

I've got two test-containers set up in my tests, LocalStack and MockServer. Both of them are configured `withReuse(true)` but for some reason MockServer container stops after tests are executed. I also noticed something else; when I start the containers in one process and run the tests in another process, LocalStack container works without any issue and stays running (next to Ryuk container) after tests are executed, but MockServer stops right after tests are done. I couldn't find anything helpful as to what could be the issue:
1. I created the container using directly GenericContainer class, same behavior.
2. I Replace MockServer with WireMock (via GenericContainer) and it behaves as expected (similar to LocalStack), stays running after tests are done in another process, and also respects the `reuse` flag.

### Relevant log output

```shell
There's not much log, only one line (apart from the startup logs):
> 19:59:50.278 [,] [docker-java-stream--1211111146] INFO  w.infra.containers.MockServer - STDOUT: 2022-11-01 18:59:50 5.14.0 INFO stopped for port: 1080
```

**Comment by eddumelendez** [comment]
@akefirad Can you share your setup for both containers, please? `MockServerContainer` doesn't have any special that prevents the `reuse` experimental feature in Testcontainers. Also, I just tried without issues, adding `testcontainers.reuse.enable=true` to `~/.testcontainers.properties`

**Comment by akefirad** [comment]
I have to clean up the project. Will do that tomorrow first thing.

**Comment by akefirad** [comment]
OK, while cleaning up the project, I found the root cause. It's a spring boot application, and as soon as I add a `MockServerClient` bean the issue appears. I suspect the reason is that `MockServerClient` implement `Closeable` which is an indication for spring to destroy the bean. I'm not 100% sure, but a quick look at the code, I don't know how to disable the behaviour in `MockServerClient`. I guess I could disable the destroy logic on Spring side, but that seems like a hack to me. IMO having the stop logic on the client side is a bit counter-intuitive, at the very least the naming is confusing, is `stop()` stopping the client or the server? If it's stopping only the server, does it not make more sense to move it out to somewhere else, or make it like an extension to the client? (Let me know if I can help with anything or if you need more information.)

**Reply by eddumelendez** [reply]
can you share an example to reproduce it? I think there are a few things to explain but would like to see it first.

**Comment by akefirad** [comment]
There you go [mock-server-demo.zip](https://github.com/testcontainers/testcontainers-java/files/9939749/mock-server-demo.zip).
Thinking more about it, not sure if you can do anything about it. It's a feature maintained by MockServer project. The behavior, if at all, should be changed on their side. 🤷 (Nevertheless it sure messes with re-use feature of TestContainers.)

**Reply by eddumelendez** [reply]
Hi, just took a quick look to the project provided but it seems to be very spring specific the issue you have. Just to be clear, about previous comment:

> I suspect the reason is that MockServerClient implement Closeable which is an indication for spring to destroy the bean

Spring is destroying the mockserver client bean, not the server. So, this is not related. When using `reuse` feature in testcontainers once the container is started manually (calling `start()` method) is doesn't stop.

If you isolate testcontainers from spring then you can confirm. It doesn't mean you can not integrate both, but there is something to take a look at it.

**Reply by akefirad** [reply]
> it seems to be very spring specific the issue you have.

Yes, as I mentioned, this is due to the way Spring manages the lifecycle of Closeable beans.

> Spring is destroying the mockserver client bean, not the server.

Spring destroys any bean that impelements `Closeable` beans (there are other scenarios, but not relevant here). In this case both the client and the server beans are managed by Spring. But Spring destroys only the client because only the client implements Closeable.

> So, this is not related.

To prove that this is indeed the root cause, you can set the bean definition to `@Bean` (i.e. remove the `destroyMethod`) for the client (this way you can see that the container gets stopped). Besides, looking at the `stop` method in the client, it's clearly stopping the server, unless I'm missing something: `getMockServerEventBus().publish(EventType.STOP);`
On that note, the fact that this has caused a confusion here proves my point: Why is the client stopping the server? (Probably a question to mock-server team, but it's affecting test containers anyway.)

**Reply by akefirad** [reply]
Just a small correction (from Spring docs):
> This 'destroy method inference' is currently limited to detecting only public, no-arg methods named 'close' or 'shutdown'. 

So it's not only about Closeable.
